src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultArr = [];
+
+  let previousState = {};
+
+  Object.assign(previousState, state);
+
+  for (const anAction of actions) {
+    for (const typeValue in anAction) {
+      switch (anAction[typeValue]) {
+        case 'addProperties':
+          Object.assign(previousState, anAction.extraData);
+          resultArr.push(Object.assign({}, previousState));
+          break;
+
+        case 'removeProperties':
+          for (const propRemove of anAction.keysToRemove) {
+            delete previousState[propRemove];
+          }
+          resultArr.push(Object.assign({}, previousState));
+          break;
+
+        case 'clear':
+          previousState = {};
+          resultArr.push(Object.assign({}, previousState));
+          break;
+      }
+    }
+  }
```

Comment:
```suggestion
  for (const action of actions) {
    switch (action.type) {
      case 'addProperties':
        Object.assign(previousState, action.extraData);
        break;

      case 'removeProperties':
        for (const key of action.keysToRemove) {
          delete previousState[key];
        }
        break;

      case 'clear':
        for (const key in previousState) {
          delete previousState[key];
        }
        break;
    }
    resultArr.push({ ... previousState });
  }
```

Reply:
Oh... Yes, you're right - your solution better and more harmonic... I will study to do it in this way. Thanks!